### DISCUSSION TITLE
<cbc:CompanyID /> is empty in XRechnung and Extended Profile

### DISCUSSION THREAD

**Discussion by fisfro** [stephanstapel/ZUGFeRD-csharp#460] [Q&A]
Hello again,

in the meantime I will pass EN16931 UBL Invoice 1.3.11 validation :-)
but  CIUS XRechnung (UBL) (val-sch.2) validation produce some headache.....
For CIUS I will get the error due an empty cbc:CompanyID /> in PartyLegalEntity in section  <cac:AccountingCustomerParty>.

But I have configured the Company ID multiple time in SetBuyer(..  like:

       desc.SetBuyer(Beleg.AnFirma, Beleg.AnPLZ, Beleg.AnOrt, Beleg.AnStrasse, GetCountryCode(Beleg.AnStaat), string.Empty, new GlobalID(GlobalIDSchemeIdentifiers.GLN, GLNBuyer), legalOrganization: new LegalOrganization(GlobalIDSchemeIdentifiers.GLN, GLNBuyer, Beleg.AnFirma)); 
  
        desc.AddBuyerTaxRegistration(Beleg.KundenUID, TaxRegistrationSchemeID.VA);//yes I know, thats needed for taxes....

Okay, to be honest , it makes not much sense to add two times the GLN id, but GlobalIDSchemeIdentifiers will not include a VAT parameter to define an UID.

What did I missed? Where to set this CompanyID? 

Yes, I know ,I could fix this problem with a replace of this empty node in xml-string, but thats not a perfect solution.

Any help will be welcome ..

Thanks in advance !

Robert

**Comment by DaveHalma** [comment]
Hi Robert, 
can you share the XRechnung you're trying to read? I think I had a similar problem and was able to fix it.
Dave

**Reply by fisfro** [reply]
> Hi Robert, can you share the XRechnung you're trying to read? I think I had a similar problem and was able to fix it. Dave

Hello Mark,

I have the problem with writing the invoice. In the invoiceDescriptor22UblWriter.cs I have found following code:
                foreach (var tax in taxRegistrations)
                {
                    Writer.WriteStartElement("cac", "PartyTaxScheme");
                    Writer.WriteElementString("cbc", "CompanyID", tax.No);
                    Writer.WriteStartElement("cac", "TaxScheme");
                    Writer.WriteElementString("cbc", "ID", UBLTaxRegistrationSchemeIDMapper.Map(tax.SchemeID));
                    Writer.WriteEndElement(); //!TaxScheme
                    Writer.WriteEndElement(); //!PartyTaxScheme
                }

So CompanyID will be writen with tax.No.

I'm using
desc.AddBuyerTaxRegistration(Beleg.KundenUID, TaxRegistrationSchemeID.VA); 

to set the tax.no and scheme. But for any reason (I'm currently not able to debug the test code of the lib) the uid will be not set. Removing this field, will lead in a positive pass for validation in all tested validators.

Till this isse is clearified I'm a fix function, which fix multiple problems  I have at the moment and also remove this empty CompanyID  : 

public static string FixUnsolvedBugs(string xmlInvoice, int BelegTypID)
{
    // Define the lines to add for TaxExemption
    string taxExemptionToAdd = @"
    <cbc:TaxExemptionReasonCode>VATEX-EU-I</cbc:TaxExemptionReasonCode>
    <cbc:TaxExemptionReason>local law</cbc:TaxExemptionReason>";

    // Split the XML into lines for processing
    List<string> updatedLines = new List<string>();
    string[] lines = xmlInvoice.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);

    bool insideTaxCategory = false; // Flag to track if we're inside the correct node

    foreach (string line in lines)
    {
        // Exclude the unwanted line "<cbc:CompanyID />"
        if (line.Contains("<cbc:CompanyID />"))
        {
            continue;
        }

        // Detect the start of a TaxCategory node
        if (line.Contains("<cac:TaxCategory>"))
        {
            insideTaxCategory = true;
        }

        // Detect the end of a TaxCategory node
        if (line.Contains("</cac:TaxCategory>"))
        {
            insideTaxCategory = false;
        }

        // Add the current line to the updated list
        updatedLines.Add(line);

        // If inside TaxCategory and line contains "<cbc:Percent>0.00</cbc:Percent>", add TaxExemption lines
        if (insideTaxCategory && line.Contains("<cbc:Percent>0.00</cbc:Percent>"))
        {
            updatedLines.AddRange(
                taxExemptionToAdd.Trim().Split(new[] { "\r\n", "\n" }, StringSplitOptions.None)
            );
        }
    }

    // Rejoin the updated lines into the final XML string
    string finalXML = string.Join("\n", updatedLines);
    // credit note fixes
    if (BelegTypID == 8)
    {
        finalXML = finalXML.Replace("<cbc:InvoiceTypeCode>381</cbc:InvoiceTypeCode>", "<cbc:CreditNoteTypeCode>381</cbc:CreditNoteTypeCode>");
        finalXML = finalXML.Replace("InvoiceLine", "CreditNoteLine");
        finalXML = finalXML.Replace("InvoicedQuantity", "CreditedQuantity");
    }
    return finalXML;
}


With this fix I will pass all validations in extended and XRechnung Mode for UBL for invoice and credit notes. Also credit notes are not working as needed for me in the libary. I.e. needed param like CreditNoteLine_ are not part of the libary- or let's say, I was not able to find the term in the lib.

CIAO Robert
